### PR TITLE
raftstore-v2: include commit index for learner snapshot

### LINKED ISSUE #15026: [Dynamic Regions] Include commit index index in snapshot message
## Development Task

See https://github.com/pingcap/tiflash/issues/7568#issuecomment-1576382311

TiFlash prefers including commit index in snapshot, so they can skip read index.

### PR DESCRIPTION
**Pull request by overvenus**
### What is changed and how it works?

Issue Number: Close #15026

What's Changed:

```commit-message
Set commit index for learner snapshots. It's needed to address
compatibility issues between v1 and v2 snapshots.

See https://github.com/pingcap/tiflash/issues/7568#issuecomment-1576382311
```

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: tests/integrations/raftstore/test_snap.rs [hunk 1]
```diff
@@ -948,3 +948,36 @@ fn check_observer(observer: &MockApplySnapshotObserver, region_id: u64, snap_pat
...
+    // compatibility issues between v1 and v2 snapshots.
+    // See https://github.com/pingcap/tiflash/issues/7568#issuecomment-1576382311
+    pd_client.must_add_peer(r, new_learner_peer(2, 2));
+    let msg = rx.recv_timeout(Duration::from_secs(5)).unwrap();
+    assert_ne!(msg.get_message().get_commit(), 0);
+
+    cluster.must_put(b"k2", b"v2");
+
+    pd_client.must_add_peer(r, new_peer(2, 2));
+    cluster.must_transfer_leader(1, new_peer(1, 1));
```
Comment (overvenus) [inline tests/integrations/raftstore/test_snap.rs:980]:
```suggestion
    cluster.must_transfer_leader(1, new_peer(2, 2));
```

File: components/raftstore-v2/src/operation/ready/mod.rs [hunk 2]
```diff
@@ -535,14 +533,21 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
...
         let to_peer_id = msg.get_to_peer().get_id();
         let to_store_id = msg.get_to_peer().get_store_id();
         if msg_type == MessageType::MsgSnapshot {
             let index = msg.get_message().get_snapshot().get_metadata().get_index();
             self.update_last_sent_snapshot_index(index);
+
+            // Set commit index for learner snapshots. It's needed to address
+            // compatibility issues between v1 and v2 snapshots.
+            // See https://github.com/pingcap/tiflash/issues/7568#issuecomment-1576382311
+            if let Some(p) = find_peer(self.region(), to_store_id) && p.get_role() == PeerRole::Learner  {
```
[the remark this answers is not in the document]
Comment (overvenus) [reply components/raftstore-v2/src/operation/ready/mod.rs]: Has moved commit index to `SnapshotMeta`. PTAL.

File: components/raftstore-v2/src/operation/ready/snapshot.rs [hunk 3]
```diff
@@ -599,6 +606,16 @@ impl<EK: KvEngine, ER: RaftEngine> Storage<EK, ER> {
                 );
                 return false;
             }
+            // Set commit index for learner snapshots. It's needed to address
+            // compatibility issues between v1 and v2 snapshots.
+            // See https://github.com/pingcap/tiflash/issues/7568#issuecomment-1576382311
+            if let Some(p) = find_peer_by_id(self.region(), to_peer_id) && p.get_role() == PeerRole::Learner  {
```
Comment (tabokie) [inline components/raftstore-v2/src/operation/ready/snapshot.rs:612]: No harm to set it unconditionally?

**Review by tabokie — APPROVED** [review]

File: components/raftstore-v2/src/operation/ready/snapshot.rs (hunk 3 above)
Comment (overvenus) [reply components/raftstore-v2/src/operation/ready/snapshot.rs:612]: Yes, it's okay. Although I prefer to set it only for learners to make snapshot clearer.

File: Cargo.toml [hunk 4]
```diff
@@ -200,8 +200,8 @@ procinfo = { git = "https://github.com/tikv/procinfo-rs", rev = "6599eb9dca74229
 # When you modify TiKV cooperatively with kvproto, this will be useful to submit the PR to TiKV and the PR to
 # kvproto at the same time.
 # After the PR to kvproto is merged, remember to comment this out and run `cargo update -p kvproto`.
-# [patch.'https://github.com/pingcap/kvproto']
-# kvproto = { git = "https://github.com/your_github_id/kvproto", branch = "your_branch" }
+[patch.'https://github.com/pingcap/kvproto']
+kvproto = { git = "https://github.com/overvenus/kvproto", branch = "snapshot-commit-index" }
```
Comment (overvenus) [inline Cargo.toml:204]: Need to merge kvproto first.

File: tests/integrations/raftstore/test_snap.rs [hunk 5]
```diff
@@ -948,3 +949,40 @@ fn check_observer(observer: &MockApplySnapshotObserver, region_id: u64, snap_pat
 
     panic!("cannot find {:?} in observer", snap_path);
 }
+
+#[test]
```
Comment (CalvinNeo) [inline tests/integrations/raftstore/test_snap.rs:953]: Is it better to put this test into test_v1_v2_mixed.rs?

**Review by CalvinNeo — APPROVED** [review]
lgtm

**Review by tonyxuqqi — APPROVED** [review]

**Pull request merged**